metaswitch/common/alarms.py

```diff
@@ -31,36 +31,281 @@
+            _log.info('Waiting for alarm manager to quiesce.')
+            self._condition.wait(4)
+
+    def _re_sync_alarms(self):
+        """Re-sync each alarm in the registry."""
+        current_alarms = self._alarm_registry.values()
+
+        # Each alarm sync may take up to 2s on failure. This could
+        # easily add up to more than the RE_SYNC_INTERVAL. In this case,
+        # an log will be written by _get_sleep_time. The expected cause
+        # of long request times is that alarm agent is not available;
+        # in this case it makes little difference which alarms we are
+        # failing to re-sync, so there is no need for timeout logic.
+        for alarm in current_alarms:
+            if self._should_terminate: # pragma: no cover
+                # There is no way of ensuring that we hit this condition.
+                break
+            alarm.re_sync()
 
-def issue_alarm(process, identifier):
+    def _update_resync_time(self):
+        """Calculate how long to sleep before the next re-sync."""
+        self._next_resync_time += 30
+        current_time = monotonic()
+        sleep_time = self._next_resync_time - current_time
+
+        if sleep_time <= 0:
+            missed_by = -sleep_time
+            _log.error('Missed alarm re-sync time by %ds', missed_by)
+            skips = ((missed_by / RE_SYNC_INTERVAL) + 1)
+            self._next_resync_time += (skips * RE_SYNC_INTERVAL)
+            sleep_time = self._next_resync_time - current_time
+
+        return sleep_time
+
+
+alarm_manager = _AlarmManager()
+
+
+class BaseAlarm(object):
+    def __init__(self, issuer, index):
+        self._clear_state = AlarmState(issuer, index, CLEARED)
+        self._last_state_raised = self._clear_state
+
+    def clear(self):
+        """Send the alarm's cleared state to the alarm agent."""
+        self._last_state_raised = self._clear_state
+        self.re_sync()
+
+    def re_sync(self):
+        """Send or re-send the alarm's state to the alarm agent."""
+        self._last_state_raised.issue()
+
+
+class Alarm(BaseAlarm):
+    """Alarm with only a single non-cleared severity.
+
+    The parameter severity should be passed a severity constant from this
+    module"""
+    def __init__(self, issuer, index, severity):
+        super(Alarm, self).__init__(issuer, index)
+        self._alarm_state = AlarmState(issuer, index, severity)
+
+    def set(self):
+        """Send the alarm's raised state to the alarm agent."""
+        self._last_state_raised = self._alarm_state
+        self.re_sync()
+
+
+class MultiSeverityAlarm(BaseAlarm):
+    """Alarm with multiple possible non-cleared severities.
+
+    The parameter severities should be passed an iterable of severity
+    constants from this module.
+    """
+    def __init__(self, issuer, index, severities):
+        super(MultiSeverityAlarm, self).__init__(issuer, index)
+        assert len(severities) >= 2, 'Must have multiple severities.'
```

Comment:
What is the behaviour on a failed assert? does just that alarm fail? or does the whole process die?